clients/venice-client/src/main/java/com/linkedin/venice/fastclient/meta/RequestBasedMetadata.java

```diff
@@ -173,20 +198,150 @@ void discoverD2Service() {
+   * warmup is on best effort basis for now. If we want to improve the long tail and potentially
+   * further improve the client startup speed by not waiting till the timeout for all the conns
+   * to be warmed up, we can think of adding more fine-grained logic wrt partitions, replicas or
+   * warmup success rate, which might succeed faster that waiting for all conns.
+   */
+  private void warmupConnectionToInstances(int currentFetchedVersion, int partitionCount)
+      throws ExecutionException, InterruptedException {
+    Set<String> newReplicasToBeWarmedUp = new HashSet<>();
+    // get all replicas
+    for (int partitionId = 0; partitionId < partitionCount; partitionId++) {
+      newReplicasToBeWarmedUp.addAll(getReplicas(currentFetchedVersion, partitionId));
+    }
+    LOGGER.info(
+        "New fetched metadata has {} instances for store {} version {}",
+        newReplicasToBeWarmedUp.size(),
+        storeName,
+        currentFetchedVersion);
+
+    // If there are warmup in progress, check its status to cancel or reuse the same warmup
+    Set<String> replicasWarmupInProgressFromPastRefresh = new HashSet<>();
+    Set<String> futuresToRemove = new HashSet<>();
+    Iterator<Map.Entry<String, CompletableFuture>> entryIterator = warmUpInstancesFutures.entrySet().iterator();
+    while (entryIterator.hasNext()) {
+      Map.Entry<String, CompletableFuture> entry = entryIterator.next();
+      String replica = entry.getKey();
+      CompletableFuture future = entry.getValue();
+      if (!newReplicasToBeWarmedUp.contains(replica)) {
+        // No need to warm up this instance anymore: cancel if already running
+        if (!future.isDone()) {
+          future.cancel(true);
+        }
+        futuresToRemove.add(replica);
+      } else {
+        // Need to be warmed up: If already in process, don't retry
+        if (!future.isDone()) {
+          // warmup started in the previous warmups for this, not restarting it: in the
+          // worst case, the first read request to this instance will try it again.
+          newReplicasToBeWarmedUp.remove(replica);
+          replicasWarmupInProgressFromPastRefresh.add(replica);
+        } else {
+          futuresToRemove.add(replica);
+        }
+      }
+    }
+    futuresToRemove.forEach(replica -> warmUpInstancesFutures.remove(replica));
+
+    if (replicasWarmupInProgressFromPastRefresh.size() != 0) {
+      LOGGER.info(
+          "H2 connection warmup to {} instances still incomplete from the previous attempt, not retrying again for these instances: {}",
+          replicasWarmupInProgressFromPastRefresh.size(),
+          String.join(",", replicasWarmupInProgressFromPastRefresh));
+    }
+
+    // Remove old instances from warmedUpInstances to keep warmedUpInstances relatively fresh
+    warmedUpInstances.retainAll(newReplicasToBeWarmedUp);
+
+    // Warmup all the new instances
+    newReplicasToBeWarmedUp.removeAll(warmedUpInstances);
```

Comment:
The logic here is a little tricky to read..
So shall we just maintain one map: `warmedUpIntanceFutures` and keep the instance entries even the future is done?
Right now, this function is maintaining two fields: ` warmedUpInstances` and `warmUpInstancesFutures`.

Reply:
cleaned it up using only `warmedUpIntanceFutures`